server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -2454,6 +2523,19 @@ public static IndexMetadata fromXContent(XContentParser parser, Map<String, Mapp
                         case KEY_STATS:
                             builder.stats(IndexMetadataStats.fromXContent(parser));
                             break;
+                        case KEY_FIELDS_FOR_MODELS:
+                            // TODO: Could probably make this more efficient
```

Comment:
As this TODO notes, this implementation is inefficient. It would be better if we could use a parser that decodes directly to a set.

Reply:
Thought on this: We could simplify and improve efficiency by deserializing to a `Map<String, List<String>>` and then update `builder.fieldsForModels` to take `Map<String, Collection<String>>`. The `processFieldsForModels` method will handle copying & converting the list to a set.

It's basically the same thing that we're doing right now, but with fewer steps and intermediate objects.

Reply:
Created a follow-up issue to optimize this parsing in a separate PR.